File: app/src/main/java/net/tlalka/android/fiszki/model/dao/WordDao.java
package net.tlalka.android.fiszki.model.dao;

import java.sql.SQLException;
import java.util.List;

import com.j256.ormlite.support.ConnectionSource;
import net.tlalka.android.fiszki.model.entities.Cluster;
import net.tlalka.android.fiszki.model.entities.Lesson;
import net.tlalka.android.fiszki.model.entities.Word;
import net.tlalka.android.fiszki.model.types.LanguageType;

public class WordDao extends AbstractDao<Word, Long> {

    public WordDao(ConnectionSource connectionSource) throws SQLException {
        super(connectionSource, Word.class);
    }

    public Word getWordBy(Cluster cluster, LanguageType languageType) throws SQLException {
        return super.queryBuilder()
                .where()
                .eq("cluster_id", cluster)
                .and()
                .eq("languageType", languageType)
                .queryForFirst();
    }

    public List<Word> getWordsBy(Lesson lesson) throws SQLException {
        return super.queryBuilder()
                .distinct()
                .where()
                .eq("lesson_id", lesson)
                .query();
    }

    public List<Word> getWordsBy(Cluster cluster) throws SQLException {
        return super.queryBuilder()
                .distinct()
                .where()
                .eq("cluster_id", cluster)
                .query();
    }

    public List<Word> getWordsBy(LanguageType languageType) throws SQLException {
        return super.queryBuilder()
                .orderBy("value", true)
                .where()
                .eq("languageType", languageType)
                .query();
    }
}

File: app/src/main/java/net/tlalka/android/fiszki/model/dao/WordDao.kt
package net.tlalka.android.fiszki.model.dao

import com.j256.ormlite.support.ConnectionSource
import net.tlalka.android.fiszki.model.entities.Cluster
import net.tlalka.android.fiszki.model.entities.Lesson
import net.tlalka.android.fiszki.model.entities.Word
import net.tlalka.android.fiszki.model.types.LanguageType
import java.sql.SQLException

class WordDao @Throws(SQLException::class)
constructor(connectionSource: ConnectionSource) : AbstractDao<Word, Long>(connectionSource, Word::class.java) {

    @Throws(SQLException::class)
    fun getWordBy(cluster: Cluster, languageType: LanguageType): Word {
        return super.queryBuilder()
                .where()
                .eq("cluster_id", cluster)
                .and()
                .eq("languageType", languageType)
                .queryForFirst()
    }

    @Throws(SQLException::class)
    fun getWordsBy(lesson: Lesson): List<Word> {
        return super.queryBuilder()
                .distinct()
                .where()
                .eq("lesson_id", lesson)
                .query()
    }

    @Throws(SQLException::class)
    fun getWordsBy(cluster: Cluster): List<Word> {
        return super.queryBuilder()
                .distinct()
                .where()
                .eq("cluster_id", cluster)
                .query()
    }

    @Throws(SQLException::class)
    fun getWordsBy(languageType: LanguageType): List<Word> {
        return super.queryBuilder()
                .orderBy("value", true)
                .where()
                .eq("languageType", languageType)
                .query()
    }
}
